### PR TITLE
added modules to support callout templates

### PR DESCRIPTION
**Pull request by techsmyth**
New entities / modules:
* CalloutFraming
* CalloutResponseDefaults
* CalloutResponsePolicy
* CalloutTemplate

Additions
* lookup storage bucket for the profile on calloutFraming and calloutTemplate
* added lookup entry for calloutTemplate
* Migration up / down (not fully tested)
* DTOs used added to validation
* samples for working with the api

Light testing is done (create, read). Additional validation needed. 

image

### CODE REVIEW

File: src/domain/collaboration/callout-framing/callout.framing.entity.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  @AfterUpdate()
+  @AfterLoad()
+  async decompressContent() {
+    if (this.content && this.content !== '') {
+      this.content = await decompressText(this.content);
+    }
+  }
+
+  @Column('longtext', { nullable: false })
+  content?: string;
```
Comment (hero101) [inline src/domain/collaboration/callout-framing/callout.framing.entity.ts]: The typeorm definition says non-nullable, but the class field is nullable.

File: src/domain/collaboration/callout-framing/callout.framing.interface.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+@ObjectType('CalloutFraming')
+export abstract class ICalloutFraming extends IAuthorizable {
+  profile!: IProfile;
+
+  @Field(() => JSON, {
+    nullable: true,
+    description: 'The whiteboard template content for this Callout Framing.',
+  })
+  content?: string;
```
Comment (hero101) [inline src/domain/collaboration/callout-framing/callout.framing.interface.ts]: I think it's a good idea to sync this field with the entity definition

File: src/domain/collaboration/callout-framing/callout.framing.service.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    private calloutFramingRepository: Repository<CalloutFraming>
+  ) {}
+
+  async createCalloutFraming(
+    calloutFramingData: CreateCalloutFramingInput
+  ): Promise<ICalloutFraming> {
+    const calloutFraming: ICalloutFraming =
+      CalloutFraming.create(calloutFramingData);
+
+    calloutFraming.authorization = new AuthorizationPolicy();
```
Comment (hero101) [inline src/domain/collaboration/callout-framing/callout.framing.service.ts:36]: what about profile

File: src/domain/collaboration/callout-response-defaults/callout.response.defaults.entity.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  @AfterUpdate()
+  @AfterLoad()
+  async decompressContent() {
+    if (this.whiteboardContent && this.whiteboardContent !== '') {
+      this.whiteboardContent = await decompressText(this.whiteboardContent);
+    }
+  }
+
+  @Column('longtext', { nullable: false })
+  whiteboardContent?: string;
```
Comment (hero101) [inline src/domain/collaboration/callout-response-defaults/callout.response.defaults.entity.ts]: The entity definition is `nullable: false`

Comment (hero101) [inline src/domain/collaboration/callout-response-defaults/callout.response.defaults.entity.ts]: The entity definition is `nullable: false`

File: src/domain/collaboration/callout-response-defaults/callout.response.defaults.service.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  async updateCalloutResponseDefaults(
+    calloutResponseDefaults: ICalloutResponseDefaults,
+    calloutResponseDefaultsData: UpdateCalloutResponseDefaultsInput
+  ): Promise<ICalloutResponseDefaults> {
+    if (calloutResponseDefaultsData.postDescription) {
+      calloutResponseDefaults.postDescription =
+        calloutResponseDefaultsData.postDescription;
+    }
+
+    return calloutResponseDefaults;
```
Comment (hero101) [inline src/domain/collaboration/callout-response-defaults/callout.response.defaults.service.ts]: what about `whiteboardContent`?

**Review by hero101 — CHANGES_REQUESTED** [review]
I have left some notes that we need to discuss and also general notes:
- passing a profile to the framing when creating a template does not feel right. Furthermore, you have to provide a display name to that profile, which feels even stranger.
- the responseDefaults input is mandatory but nothing in IS. I am guessing it needs to be also optional or some sort of validation to provide at least one or only one
- I believe the updating API can be improved. Updating the framing of a template requires it's ID, but it should be already known by the backend so it's not really required.

The create and delete work fine.

File: src/domain/collaboration/callout-framing/callout.framing.entity.ts (hunk 1 above)
Comment (techsmyth) [reply src/domain/collaboration/callout-framing/callout.framing.entity.ts]: nullable should be true

File: src/domain/collaboration/callout-framing/callout.framing.service.ts (hunk 3 above)
Comment (techsmyth) [reply src/domain/collaboration/callout-framing/callout.framing.service.ts:36]: yikes

File: src/domain/collaboration/callout-framing/callout.framing.entity.ts (hunk 1 above)
Comment (techsmyth) [reply src/domain/collaboration/callout-framing/callout.framing.entity.ts]: field to be called whiteboardContent

File: src/domain/collaboration/callout-response-defaults/callout.response.defaults.entity.ts (hunk 4 above)
Comment (techsmyth) [reply src/domain/collaboration/callout-response-defaults/callout.response.defaults.entity.ts]: nullable: true

File: src/domain/collaboration/callout-response-defaults/callout.response.defaults.service.ts (hunk 5 above)
Comment (techsmyth) [reply src/domain/collaboration/callout-response-defaults/callout.response.defaults.service.ts]: good catch

**Comment by techsmyth** [comment]
@hero101 please rename Response to be Contribution

**Review by hero101 — APPROVED** [review]

**Pull request merged**